File: address/go/multiAddress_test.go
package address_test

import (
	"testing"

	"github.com/stretchr/testify/require"

	"multi_address"
)

const (
	raw      = "70586beeb7b7aa2e7966df9c8493c6cbfd75c625"
	checksum = "0x70586BeEB7b7Aa2e7966DF9c8493C6CbFd75C625"
	xko      = "XKO70586BeEB7b7Aa2e7966DF9c8493C6CbFd75C625"
)

func TestToEvmAddress(t *testing.T) {
	tests := []struct {
		name string
		in   string
		want string
	}{
		{"0x prefixed", "0x" + raw, checksum},
		{"bare hex", raw, checksum},
		{"XKO prefixed", "XKO" + raw, checksum},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got, err := address.ToEvmAddress(tt.in)
			require.NoError(t, err)
			require.Equal(t, tt.want, got)
		})
	}

	_, err := address.ToEvmAddress("0x1234")
	require.ErrorContains(t, err, "invalid address length")

	_, err = address.ToEvmAddress("0xzbcdefabcdefabcdefabcdefabcdefabcdefabcd")
	require.ErrorContains(t, err, "invalid hex characters")
}

func TestFromEvmAddress(t *testing.T) {
	tests := []struct {
		name string
		in   string
		want string
	}{
		{"0x prefixed", "0x" + raw, xko},
		{"bare hex", raw, xko},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got, err := address.FromEvmAddress(tt.in)
			require.NoError(t, err)
			require.Equal(t, tt.want, got)
		})
	}

	_, err := address.FromEvmAddress("0xZZZDEFABCDEFABCDEFABCDEFABCDEFABCDEFABCD")
	require.ErrorContains(t, err, "invalid hex characters")
}

File: address/js/multiAddress.test.js
const test = require("node:test");
const assert = require("node:assert");
const { toEvmAddress, fromEvmAddress } = require("./multiAddress");

const rawAddress = "70586beeb7b7aa2e7966df9c8493c6cbfd75c625";
const checksummedAddress = "70586BeEB7b7Aa2e7966DF9c8493C6CbFd75C625";
const hexPrefixChecksummedAddress = "0x70586BeEB7b7Aa2e7966DF9c8493C6CbFd75C625";

// test toEvmAddress
test("normalizes already 0x-prefixed hex", () => {
  const input = "0x"+rawAddress;
  console.log("converted to", toEvmAddress(input));
  assert.strictEqual(
    toEvmAddress(input),
    hexPrefixChecksummedAddress
  );
});

test("normalizes bare hex", () => {
  const input = rawAddress;
  assert.strictEqual(
    toEvmAddress(input),
    hexPrefixChecksummedAddress
  );
});

test("normalizes XKO-prefixed hex", () => {
  const input = "XKO"+rawAddress;
  assert.strictEqual(
    toEvmAddress(input),
    hexPrefixChecksummedAddress
  );
});

test("rejects wrong length", () => {
  assert.throws(
    () => toEvmAddress("0x1234"),
    /Invalid address length/
  );
});

test("rejects invalid chars", () => {
  assert.throws(
    () => toEvmAddress("0xzbcdefabcdefabcdefabcdefabcdefabcdefabcd"),
    /Invalid hex characters/
  );
});

test("rejects non-string input", () => {
  assert.throws(() => toEvmAddress(123), /Address must be a string/);
});

// test fromEvmAddress
test("converts 0x prefixed EVM address back to XKO", () => {
  assert.strictEqual(
    fromEvmAddress("0x"+rawAddress),
    "XKO"+checksummedAddress
  );
});

test("converts EVM address back to XKO", () => {
  assert.strictEqual(
    fromEvmAddress(rawAddress),
    "XKO"+checksummedAddress
  );
});

test("fromEvmAddress rejects non-string input", () => {
  assert.throws(() => fromEvmAddress(123), /Address must be a string/);
});

test("fromEvmAddress rejects invalid hex", () => {
  assert.throws(
    () => fromEvmAddress("0xZZZDEFABCDEFABCDEFABCDEFABCDEFABCDEFABCD"),
    /Invalid hex characters/
  );
});